tango_simlib/compat.py

```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf8 -*-
+# vim:fileencoding=utf8 ai ts=4 sts=4 et sw=4
+# Copyright 2019 National Research Foundation (South African Radio Astronomy Observatory)
+# BSD license - see LICENSE for details
+
+"""Utilities for dealing with Python 2 and 3 compatibility."""
+
+from __future__ import absolute_import, division, print_function
+from future import standard_library
+
+standard_library.install_aliases()  # noqa: E402
+
+import future
+
+if future.utils.PY2:
+
+    def ensure_native_str(value):
+        """Coerce unicode string or bytes to native string type (UTF-8 encoding)."""
```

Comment:
I copied the method name and signature from https://github.com/ska-sa/katcp-python/blob/master/katcp/compat.py .
I plan to change the name to something more appropriate(perhaps `unicode_to_ascii` when the functionality of what is required from other reviewers are given.

Reply:
Okay thanks, the reason i was asking for that improvement was that the function behaves differently provided the given criteria i thought it would be very informative to know the difference of the functions through the docstring but seems okay now.

Reply:
Yes, It does not `ensure_native_str` anymore but there is some conversion from unicode strings to a safe ascii version. 
See https://github.com/ska-sa/tango-simlib/blob/9b9ccb7df9477d92fe3b8d1ae2a5b12c79fe2ab0/tango_simlib/utilities/sim_xmi_parser.py#L136 

In the python 3 version we might need to do something else or remove the above line of code as we can parse and process unicode.

Reply:
Looking at the usage of this function, it seems like `ensure_byte_string` or `ensure_ascii_byte_string` would be better names.  And if that is the usage, then, the implementations here aren't quite right.  Py2 is OK, but Py3 is still returning Unicode string.

What we need is more like `tobytes` - https://python-future.org/reference.html#future.utils.tobytes.  It uses `latin-1` encoding instead of `ascii`, but that should be OK for Tango.  However, it doesn't enable `"replace"`, so we need something custom.  Our own function should use `"ascii"` and "`"replace"`.  And raise errors if the input type is wrong.

Reply:
after changing the method name the tests broke and when I reverted the changes the tests have now become flaky. I am looking into these changes.

Reply:
I think the flakiness is unrelated to these name changes.  I  got 4/4 passes (runs 12 to 15) after changing the test runs from parallel to serial (using Jenkins replay).  I don't think we should run these tests in different Python versions simultaneously. We are starting devices with the same name, and Tango won't allow more than one to run (when using the Tango DB).  Also, if devices are run via the test context, the same port number is always used (8888), which could also lead to failure.

What is a little strange is that the recent failures on Py3 have `ImportError: No module named tango_simlib.tango_sim_generator` errors which is a bit strange. Looking at the timestamps of these runs, the Py2 installation is typically still busy while Py3 has already installed and started running the tests.  I don't know why the Py2 installation process might affect the Py3 tests though.

Reply:
Thanks for valued input I was not aware of the inner workings and setup of these tango-simlib tests.